<a href="https://colab.research.google.com/github/KirillSmirnov2/InnopolisStudentPublicRepository/blob/main/HW3(dataset_from_HW2)_SmirnovKK.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [288]:
import pandas as pd
import time
df1=pd.read_csv('anime.csv')

In [106]:
# @title Используем команды '.info()'  и  '.describe()' для получение информации о датасете.
df1.info()
df1.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


anime_id        rating       members
count  12294.000000  12064.000000  1.229400e+04
mean   14058.221653      6.473902  1.807134e+04
std    11455.294701      1.026746  5.482068e+04
min        1.000000      1.670000  5.000000e+00
25%     3484.250000      5.880000  2.250000e+02
50%    10260.500000      6.570000  1.550000e+03
75%    24794.500000      7.180000  9.437000e+03
max    34527.000000     10.000000  1.013917e+06

In [107]:
# @title Определим общий список жанров в датасете
n_rows=len(df1['genre'])
print('число строк:  ',n_rows)
genre_list=[]
count_genres=0
for i in range(n_rows): #n_rows#
  S1=[]
  if type(df1['genre'].loc[i]) == str:
    S1=df1['genre'].loc[i].split(',')
    P=[]
    for j in S1:
      P=j.replace(' ','')
      if genre_list.count(P)==0:
        genre_list.append(P)
  else:
    pass
print('список жанров:  ',genre_list)
print('число жанров:  ',len(genre_list))







число строк:   12294
список жанров:   ['Drama', 'Romance', 'School', 'Supernatural', 'Action', 'Adventure', 'Fantasy', 'Magic', 'Military', 'Shounen', 'Comedy', 'Historical', 'Parody', 'Samurai', 'Sci-Fi', 'Thriller', 'Sports', 'SuperPower', 'Space', 'SliceofLife', 'Mecha', 'Music', 'Mystery', 'Seinen', 'MartialArts', 'Vampire', 'Shoujo', 'Horror', 'Police', 'Psychological', 'Demons', 'Ecchi', 'Josei', 'ShounenAi', 'Game', 'Dementia', 'Harem', 'Cars', 'Kids', 'ShoujoAi', 'Hentai', 'Yaoi', 'Yuri']
число жанров:   43


**Задание 3 на основе данных из Задания 2**

**Продолжить исследование датасета, который был использован в прошлом домашнем задании. Выполнить следующие действия: **

- проверить датасет на наличие пропусков, если они есть, вывести строки с пропусками на экран;
- провести оценку корреляции между столбцами датафрейма и сделать выводы;
- произвольно применить функции сложения, нахождения среднего значения, медианы, минимума и максимума, стандартного отклонения для случайной выборки из датасета;
- применить операции сравнения.


In [108]:
# @title Проверка dataset на наличие пропусков заполнения столбцов "жанра" и "типа"
df2=df1.isnull()
df_NaN_Type=df1.loc[df2['type'] == True]
df_NaN_Genre=df1.loc[df2['genre'] == True]
print('число строк без данных по типу :  ', len(df_NaN_Type) )
print('число строк без данных по жанру :  ', len(df_NaN_Genre))
print('Общее число строк  данных :  ', len(df1))
print('==============================================')
print('строки с пропусками информации по жанру :  ')
print(df_NaN_Genre[['name' ,'genre','type']])
print('==============================================')
print('строки с пропусками информации по типу :  ')
print(df_NaN_Type[['name' ,'genre','type']])
print('==============================================')

число строк без данных по типу :   25
число строк без данных по жанру :   62
Общее число строк  данных :   12294
строки с пропусками информации по жанру :  
                                                    name genre     type
2844           IS: Infinite Stratos 2 - Infinite Wedding   NaN  Special
3541                                       ViVid Strike!   NaN       TV
6040                                   Metropolis (2009)   NaN    Movie
6646                                        Match Shoujo   NaN      ONA
7018                                    Katsudou Shashin   NaN    Movie
...                                                  ...   ...      ...
11070                                    Seikaisuru Kado   NaN      NaN
11086  Tamagotchi Movie: Tanpen Himitsu no Otodoke Da...   NaN    Movie
11097                             Tsukipro The Animation   NaN       TV
11112                                  Yuyushiki Special   NaN  Special
11113                                      Zunda Ho

In [109]:
# @title Провести оценку корреляции между столбцами датафрейма и сделать выводы: Проверим корреляцию между рейтингом[rating] и количеством участников сообщества[members].
df_NaN_ratings=df1.loc[df2['rating'] == True]

print(df_NaN_ratings[['anime_id' ,'rating','members']])
print('число строк без данных по рейтингу [rating] :  ', len(df_NaN_ratings) )
print('==============================================')

df_NaN_members=df1.loc[df2['members'] == True]
print(df_NaN_members[['anime_id' ,'rating','members']])
print('число строк без данных по числу участников [members] :  ', len(df_NaN_members) )
print('==============================================')
df_ratings_OK=df1.loc[df2['rating'] == False]
df_corr=df_ratings_OK.corr()
print(len(df_ratings_OK))
print('проверка строк:  ',n_rows,' - ',len(df_ratings_OK),'=',(n_rows - len(df_ratings_OK)), '|  ' , len(df_NaN_ratings))
print(df_corr)


       anime_id  rating  members
8968      34502     NaN       32
9657      34309     NaN      129
10896     34096     NaN    13383
10897     34134     NaN    90706
10898     30484     NaN    60999
...         ...     ...      ...
12274     34492     NaN      392
12279     34491     NaN       79
12280     34312     NaN      240
12282     34388     NaN      195
12285     34399     NaN      485

[230 rows x 3 columns]
число строк без данных по рейтингу [rating] :   230
Empty DataFrame
Columns: [anime_id, rating, members]
Index: []
число строк без данных по числу участников [members] :   0
12064
проверка строк:   12294  -  12064 = 230 |   230
          anime_id    rating   members
anime_id  1.000000 -0.284625 -0.075500
rating   -0.284625  1.000000  0.387979
members  -0.075500  0.387979  1.000000


<ipython-input-109-b95431fff4bb>:13: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_corr=df_ratings_OK.corr()


Вывод: Слабая корреляция (r=0.387979) между числом участников и рейтингом. Рейтинг мало зависит от числа участкников сообщества.




**Произвольно применить функции сложения, нахождения среднего значения, медианы, минимума и максимума, стандартного отклонения для случайной выборки из датасета;**

Определеим число участников сообществ связанных с жанрами:
[Shounen]
[Hentai]
[Mecha]


__ --------------------------------------
Загрузим файл с записями оценок [rating.csv]
__ --------------------------------------

In [110]:
dfT=pd.read_csv('rating.csv')
dfT.info()
dfT.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7813737 entries, 0 to 7813736
Data columns (total 3 columns):
 #   Column    Dtype
---  ------    -----
 0   user_id   int64
 1   anime_id  int64
 2   rating    int64
dtypes: int64(3)
memory usage: 178.8 MB


user_id      anime_id        rating
count  7.813737e+06  7.813737e+06  7.813737e+06
mean   3.672796e+04  8.909072e+03  6.144030e+00
std    2.099795e+04  8.883950e+03  3.727800e+00
min    1.000000e+00  1.000000e+00 -1.000000e+00
25%    1.897400e+04  1.240000e+03  6.000000e+00
50%    3.679100e+04  6.213000e+03  7.000000e+00
75%    5.475700e+04  1.409300e+04  9.000000e+00
max    7.351600e+04  3.451900e+04  1.000000e+01

Теперь есть два файла по которым можно выполнять анализ

Определим списки [anime_id] по каждому жанру:  [Shounen]; [Hentai]; [Mecha]
Определим число участников сообществ [members] по [anime_id] каждому жанру по данным из [anime.csv]

определим число участников сообществ по оценкам [rating]  от [user_id] из [rating.csv]   по определенным [anime_id] (по данным из [anime.csv])


In [146]:
def Count_Members(fls,AL_G, STR_PRT, NM):
  T0 = time.time()
  Kd=0
  Kd1=0
  print('[', end='')
  for j in range(len(AL_G)):
    if fls == 1:
      NM=NM+df1['members'].loc[df1['anime_id'] == AL_G[j]].values
    elif fls == 2:
      NM=NM+len(dfT.loc[dfT['anime_id'] == AL_G[j]])
    else:
      pass
    Kd1=j/len(AL_G)*10 // 1
    if Kd1>Kd:
      Kd = Kd1
      print('-', end='')
  T1 = time.time()
  print('-]')
  print(STR_PRT,' - ОК; ProcessTime: ', (T1-T0), ' sec')
  SR=NM
  return SR


In [147]:
AnimeIdList_genre_Shounen=[]
AnimeIdList_genre_Hentai=[]
AnimeIdList_genre_Mecha=[]
for i in range(n_rows):
  S1=[]
  if type(df1['genre'].loc[i]) == str:
    S1=df1['genre'].loc[i].split(',')
    P=[]
    for j in S1:
      P=j.replace(' ','')
      if P=='Shounen':
        AnimeIdList_genre_Shounen.append(df1['anime_id'].loc[i])
      else:
        if P=='Hentai':
          AnimeIdList_genre_Hentai.append(df1['anime_id'].loc[i])
        else:
          if P=='Mecha':
            AnimeIdList_genre_Mecha.append(df1['anime_id'].loc[i])
          else:
            pass
  else:
    pass

print('==============================================')
print('количество аниме жанра [Shounen]:   ',len(AnimeIdList_genre_Shounen))
print('количество аниме жанра [Hentai]:   ',len(AnimeIdList_genre_Hentai))
print('количество аниме жанра [Mecha]:   ',len(AnimeIdList_genre_Mecha))
print('==============================================')
NM_v1_Shounen, NM_v2_Shounen, NM_v1_Hentai, NM_v2_Hentai, NM_v1_Mecha, NM_v2_Mecha=0, 0, 0, 0, 0, 0

#efev=df1['members'].loc[df1['anime_id'] == 820]
#print(efev.values)
  #return

NM_v1_Shounen = Count_Members(1,AnimeIdList_genre_Shounen,'NM_v1_Shounen',NM_v1_Shounen)
NM_v1_Hentai=Count_Members(1,AnimeIdList_genre_Hentai,'NM_v1_Hentai',NM_v1_Hentai)
NM_v1_Mecha=Count_Members(1,AnimeIdList_genre_Mecha,'NM_v1_Mecha',NM_v1_Mecha)
print('==============================================')
NM_v2_Shounen=Count_Members(2,AnimeIdList_genre_Shounen,'NM_v2_Shounen',NM_v2_Shounen)
NM_v2_Hentai=Count_Members(2,AnimeIdList_genre_Hentai,'NM_v2_Hentai',NM_v2_Hentai)
NM_v2_Mecha=Count_Members(2,AnimeIdList_genre_Mecha,'NM_v2_Mecha',NM_v2_Mecha)
print('==============================================')

print('================== по файлу anime.csv =======================')
print('количество участников сообществ [members] по [anime_id] жанра [Shounen]:   ', NM_v1_Shounen)
print('количество участников сообществ [members] по [anime_id] жанра [Hentai]:   ', NM_v1_Hentai)
print('количество участников сообществ [members] по [anime_id] жанра [Mecha]:   ', NM_v1_Mecha)
print('==============================================')
print('================== по файлу rating.csv =======================')
print('количество участников сообществ [members] по [anime_id] жанра [Shounen]:   ', NM_v2_Shounen)
print('количество участников сообществ [members] по [anime_id] жанра [Hentai]:   ', NM_v2_Hentai)
print('количество участников сообществ [members] по [anime_id] жанра [Mecha]:   ', NM_v2_Mecha)
print('==============================================')



число аниме жанра [Shounen]:    1712
число аниме жанра [Hentai]:    1141
число аниме жанра [Mecha]:    944
[----------]
NM_v1_Shounen  - ОК; ProcessTime:  0.6012957096099854  sec
[----------]
NM_v1_Hentai  - ОК; ProcessTime:  0.4955778121948242  sec
[----------]
NM_v1_Mecha  - ОК; ProcessTime:  0.32584261894226074  sec
[----------]
NM_v2_Shounen  - ОК; ProcessTime:  20.461875915527344  sec
[----------]
NM_v2_Hentai  - ОК; ProcessTime:  18.953883409500122  sec
[----------]
NM_v2_Mecha  - ОК; ProcessTime:  11.552688360214233  sec
================== по файлу anime.csv =======================
число участников сообществ [members] по [anime_id] жанра [Shounen]:    [54115368]
число участников сообществ [members] по [anime_id] жанра [Hentai]:    [3030320]
число участников сообществ [members] по [anime_id] жанра [Mecha]:    [14983181]
================== по файлу rating.csv =======================
число участников сообществ [members] по [anime_id] жанра [Shounen]:    1882227
число участников соо

In [ ]:
dfT.loc[dfT['anime_id'] == 820]



In [264]:

D2_Cnt=[]
D2_MV=[]
D2_Max=[]
D2_Min=[]
D2_Med=[]
D2_SD=[]

dfTrtOK=dfT.loc[dfT['rating'] != -1]

T0 = time.time()
Kd=0
Kd1=0
ID_list=[]
#print('[', end='')
sss=0
for i in range(n_rows):  #n_rows
  ID=df1['anime_id'].loc[i]
  STRR=df1['name'][df1['anime_id'] == ID].values[0]
  ID_list.append(STRR)

  if sss>9:
    print('',end='\n')
    sss=0
  if i%50 == 0 and sss<10:
     print(i,' ', end='')
     sss=sss+1

  B1=len(dfTrtOK.loc[dfTrtOK['anime_id'] == ID])
  D2_Cnt.append(B1)
  if B1 !=0:

    D2_MV.append(dfTrtOK['rating'].loc[dfTrtOK['anime_id'] == ID].mean())
    D2_Max.append(max(dfTrtOK['rating'].loc[dfTrtOK['anime_id'] == ID]))
    D2_Min.append(min(dfTrtOK['rating'].loc[dfTrtOK['anime_id'] == ID]))
    D2_Med.append(dfTrtOK['rating'].loc[dfTrtOK['anime_id'] == ID].median())
    D2_SD.append(dfTrtOK['rating'].loc[dfTrtOK['anime_id'] == ID].std())
  else:
    D2_MV.append('')
    D2_Max.append('')
    D2_Min.append('')
    D2_Med.append('')
    D2_SD.append('')

  #Kd1=(i/n_rows)*10 // 1
  #if Kd1>Kd:
    Kd = Kd1
    #print('-', end='')
  #T1 = time.time()
print('-]')





0  50  100  150  200  250  300  350  400  450  
500  550  600  650  700  750  800  850  900  950  
1000  1050  1100  1150  1200  1250  1300  1350  1400  1450  
1500  1550  1600  1650  1700  1750  1800  1850  1900  1950  
2000  2050  2100  2150  2200  2250  2300  2350  2400  2450  
2500  2550  2600  2650  2700  2750  2800  2850  2900  2950  
3000  3050  3100  3150  3200  3250  3300  3350  3400  3450  
3500  3550  3600  3650  3700  3750  3800  3850  3900  3950  
4000  4050  4100  4150  4200  4250  4300  4350  4400  4450  
4500  4550  4600  4650  4700  4750  4800  4850  4900  4950  
5000  5050  5100  5150  5200  5250  5300  5350  5400  5450  
5500  5550  5600  5650  5700  5750  5800  5850  5900  5950  
6000  6050  6100  6150  6200  6250  6300  6350  6400  6450  
6500  6550  6600  6650  6700  6750  6800  6850  6900  6950  
7000  7050  7100  7150  7200  7250  7300  7350  7400  7450  
7500  7550  7600  7650  7700  7750  7800  7850  7900  7950  
8000  8050  8100  8150  8200  8250  8300  8350 

In [285]:
print('name ', ID_list)
print('', df1['anime_id'])
print('D2_Cnt ', len(D2_Cnt) ,D2_Cnt)
print('D2_MV ', len(D2_MV), D2_MV)
print('D2_Max ', len(D2_Max) ,D2_Max)
print('D2_Min ', len(D2_Min) ,D2_Min)
print('D2_Med ', len(D2_Med) ,D2_Med)
print('D2_SD ', len(D2_SD),D2_SD)
summa=0
for j in D2_Cnt:
  summa+=j

#print(summa)

name  ['Kimi no Na wa.', 'Fullmetal Alchemist: Brotherhood', 'Gintama°', 'Steins;Gate', 'Gintama&#039;', 'Haikyuu!!: Karasuno Koukou VS Shiratorizawa Gakuen Koukou', 'Hunter x Hunter (2011)', 'Ginga Eiyuu Densetsu', 'Gintama Movie: Kanketsu-hen - Yorozuya yo Eien Nare', 'Gintama&#039;: Enchousen', 'Clannad: After Story', 'Koe no Katachi', 'Gintama', 'Code Geass: Hangyaku no Lelouch R2', 'Haikyuu!! Second Season', 'Sen to Chihiro no Kamikakushi', 'Shigatsu wa Kimi no Uso', 'Mushishi Zoku Shou 2nd Season', 'Ookami Kodomo no Ame to Yuki', 'Code Geass: Hangyaku no Lelouch', 'Hajime no Ippo', 'Rurouni Kenshin: Meiji Kenkaku Romantan - Tsuioku-hen', 'Cowboy Bebop', 'One Punch Man', 'Mononoke Hime', 'Suzumiya Haruhi no Shoushitsu', 'Monogatari Series: Second Season', 'Mushishi Zoku Shou', 'Mushishi', 'Tengen Toppa Gurren Lagann', 'Great Teacher Onizuka', 'Natsume Yuujinchou Go', 'Hajime no Ippo: New Challenger', 'Mushishi Zoku Shou: Suzu no Shizuku', 'Natsume Yuujinchou Shi', 'Howl no Ugoku S

In [303]:
#dfT.loc[dfT['rating'] != -1]
#df1.iloc[df1['anime_id'] == 830]['name']
#STRR=df1['name'][df1['anime_id'] == 34502].values[0]
#print(STRR)

#print(df1_plus)
sera1=pd.Series(D2_Cnt,name='D2_Cnt')
sera2=pd.Series(D2_MV,name='D2_MV')
sera3=pd.Series(D2_Max,name='D2_Max')
sera4=pd.Series(D2_Min,name='D2_Min')
sera5=pd.Series(D2_Med,name='D2_Med')
sera6=pd.Series(D2_SD,name='D2_SD')

df_e = pd.concat([df1, sera1.to_frame(),sera2.to_frame(),sera3.to_frame(),sera4.to_frame(),sera5.to_frame(),sera6.to_frame()], axis = 1 ) #, ignore_index=True  #columns=['D2_Cnt','D2_MV','D2_Max','D2_Min','D2_Med','D2_SD']

print(df_e)



       anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      

In [301]:
dfT.loc[dfT['anime_id'] == 26081]

user_id  anime_id  rating
2946544    27364     26081      -1
5105671    48766     26081      -1